File: LeetCode_Algorithm/111(Minimum Depth of Binary Tree)/111.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
from typing import Optional
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        # sol1 547ms 94%
        # O(n) O(2^d)
        # from collections import deque
        # if not root:
        #     return 0
        # queue = deque([(root, 1)])
        # while queue:
        #     node, n = queue.popleft()
        #     if not node:
        #         continue
        #     if not node.left and not node.right:
        #         return n
        #     else:
        #         queue.append((node.left, n+1))
        #         queue.append((node.right, n+1))

        # sol2 533ms 82% 50.5MB 70%
        if not root:
            return 0
        self.ans = 9999999
        self.solve(root, 1)
        return self.ans
    def solve(self, node, depth):
        if not node or depth >= self.ans:
            return ;
        if not node.left and not node.right:
            self.ans = min(self.ans, depth)
            return ;
        self.solve(node.left, depth+1)
        self.solve(node.right, depth+1)
        
                

File: LeetCode_Algorithm/111(Minimum Depth of Binary Tree)/111.js
var solve2 = (node, depth) => {
  // sol2 261ms 43% 83MB 70%
  if (!node || depth >= ans) 
    return 
  if (!node.left && !node.right) {
    ans = Math.min(ans, depth)
    return ;
  }
  solve2(node.left, depth+1)
  solve2(node.right, depth+1)
}
var minDepth = function(root) {
  // sol1 402ms 18%
  // O(n) O(2^d)
  if (!root) {
    return 0
  }  
  let queue = [[root, 1]]
  while (queue.length) {
    [node, n]= queue.shift()
    if (!node) {
      continue
    } else if (!node.left && !node.right) {
      return n
    } else {
      queue.push([node.left, n+1])
      queue.push([node.right, n+1])
    }
  }
};